std-bits/table/src/main/java/org/enso/table/data/column/operation/cast/CastOperation.java

```diff
@@ -56,4 +65,422 @@ private static StorageConverter<?> fromStorageType(StorageType<?> storageType) {
+
+    return accumulator.resolveType();
+  }
+
+  private static StorageType<?> inferBigIntegerType(
+      ColumnStorage<?> columnStorage, PreciseTypeOptions options) {
+    if (!(columnStorage.getType() instanceof BigIntegerType bigIntegerType)) {
+      throw new IllegalArgumentException(
+          "Cannot infer integer type from non-integer storage: " + columnStorage.getType());
+    }
+
+    // Build the min and max of values in the column.
+    try {
+      var accumulator = new LongAccumulator();
+      StorageIterators.buildOverStorage(
+          bigIntegerType.asTypedStorage(columnStorage),
+          false,
+          bigIntegerType.makeBuilder(0, BlackholeProblemAggregator.INSTANCE),
+          (builder, index, item) -> {
+            if (item == null) {
+              return;
+            }
+            accumulator.accumulate(item.longValueExact(), false);
+          });
+
+      return options.shrinkIntegers()
+          ? accumulator.resolveType()
+          : (accumulator.getCount() == 0 ? bigIntegerType : IntegerType.INT_64);
+    } catch (ArithmeticException e) {
+      // If we cannot convert the value to long, we return the original type.
+      return bigIntegerType;
+    }
+  }
+
+  private static StorageType<?> inferFloatType(
+      ColumnStorage<?> columnStorage, PreciseTypeOptions options) {
+    if (!options.wholeFloatsBecomeIntegers()) {
+      return columnStorage.getType();
+    }
+
+    if (!(columnStorage.getType() instanceof FloatType floatType)) {
+      throw new IllegalArgumentException(
+          "Cannot infer float type from non-integer storage: " + columnStorage.getType());
+    }
+
+    // Build the min and max of values in the column.
+    try {
+      var accumulator = new LongAccumulator();
+      StorageIterators.buildOverDoubleStorage(
+          floatType.asTypedStorage(columnStorage),
+          false,
+          floatType.makeBuilder(0, BlackholeProblemAggregator.INSTANCE),
+          (builder, index, item, isNothing) -> {
+            if (isNothing) {
+              return;
+            }
+            if (item % 1 != 0 || !IntegerType.INT_64.fits(item)) {
+              throw new ArithmeticException(
+                  "Value is not a whole number or doesn't fit in a long: " + item);
+            }
+            accumulator.accumulate((long) item, isNothing);
+          });
+
+      return accumulator.getCount() == 0
+          ? floatType
+          : (options.shrinkIntegers() ? accumulator.resolveType() : IntegerType.INT_64);
+    } catch (ArithmeticException e) {
+      // If we cannot convert the value to long, we return the original type.
+      return floatType;
+    }
+  }
+
+  private static class BigDecimalAccumulator extends LongAccumulator {
+    private boolean overflowed = false;
+
+    public void accumulate(BigDecimal item) {
+      if (item == null) {
+        return;
+      }
+
+      var bigIntegerValue = item.toBigIntegerExact();
```

Comment:
I think I don't understand what happens here if `item % 1.0 != 0.0`?

Or if we can assume `item % 1.0 != 0.0`, where does this assumption come from?

If we can assume that, I'd add an assertion + a comment why.

But tbh I'm not sure why we are assuming that? Are we not missing a `%` check?

Reply:
Ah! I did not notice you had a `catch` there.

I knew it will throw but somehow didn't notice it is handled. Looks good then.